### PR TITLE
Feature 12898 Support SR for COOP

### PR DESCRIPTION
**Pull request by ritvick**
*Issue #:* bcgov/entity/issues/12898

*Description of changes:*
Support SR for COOP

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/services/filings/validations/validation.py [hunk 1]
```diff
@@ -88,6 +89,18 @@ def validate(business: Business, filing_json: Dict) -> Error:  # pylint: disable
...
+        if err:
+            return err
+
+        if 'changeOfName' in filing_json['filing'].keys():
+            err = con_validate(business, filing_json)
+        if 'alteration' in filing_json['filing'].keys():
+            err = alteration_validate(business, filing_json)
+
+        if err:
+            return err
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/filings/validations/validation.py:103]: Why do you need this elif? below else loop through these validation anyway

Comment (ritvick) [reply legal-api/src/legal_api/services/filings/validations/validation.py]: I missed one logic here, just added now. If it is a SR for COOP then we either need CoN or Alteration along with it.

File: legal-api/src/legal_api/services/filings/validations/validation.py [hunk 2]
```diff
@@ -88,6 +89,26 @@ def validate(business: Business, filing_json: Dict) -> Error:  # pylint: disable
...
+            err = con_validate(business, filing_json)
+        if 'alteration' in filing_json['filing'].keys():
+            either_con_or_alteration_flag = True
+            err = alteration_validate(business, filing_json)
+
+        if err:
+            return err
+
+        if not either_con_or_alteration_flag:
+            err = Error(HTTPStatus.BAD_REQUEST, [{'error': babel('Either Change of Name or Alteration is required.'),
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/filings/validations/validation.py:110]: I am not sure about this. @thorwolpert can you confirm this is how it should work for COOP.

Comment (argush3) [reply legal-api/src/legal_api/services/filings/validations/validation.py:110]: I believe this is correct.  

The four possible scenarios are as follows.  scenario 1 is basically irrelevant as it doesn't require a filing.
1. address change is made and is instantly saved -> no filing is required
2. only articles changes -> special resolution + alteration
3. CoN + articles changes -> special resolution + CoN + alteration
4. CoN -> special resolution + CoN

I still think some additional validation will need to happen around the rules and memorandum.  For example, key and filename are both required for rules and memorandum doc updates.  If only one of key or filename are provided, we will need to provide a validation error.

Comment (ritvick) [reply legal-api/src/legal_api/services/filings/validations/validation.py:110]: Added the new validations

File: legal-api/src/legal_api/services/filings/validations/alteration.py [hunk 3]
```diff
@@ -120,3 +122,31 @@ def type_change_validation(filing):
...
+
+def rules_change_validation(filing):
+    """Validate rules change."""
+    msg = []
+    rules_file_key: Final = get_str(filing, '/filing/alteration/rulesFileKey')
+    rules_file_name: Final = get_str(filing, '/filing/alteration/rulesFileName')
+
+    if rules_file_key or rules_file_name:
+        if not (rules_file_key and rules_file_name):
+            msg.append({'error': babel('Both rulesFileKey and rulesFileName should be privided'),
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/alteration.py:135]: typo on "provided"

File: legal-api/src/legal_api/services/filings/validations/alteration.py [hunk 4]
```diff
@@ -120,3 +122,31 @@ def type_change_validation(filing):
...
+
+def memorandum_change_validation(filing):
+    """Validate memorandum change."""
+    msg = []
+    memorandum_file_key: Final = get_str(filing, '/filing/alteration/memorandumFileKey')
+    memorandum_file_name: Final = get_str(filing, '/filing/alteration/memorandumFileName')
+
+    if memorandum_file_key or memorandum_file_name:
+        if not (memorandum_file_key and memorandum_file_name):
+            msg.append({'error': babel('Both memorandumFileKey and memorandumFileName should be privided'),
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/alteration.py:149]: type on "provided"

File: legal-api/tests/unit/services/filings/validations/test_alteration.py [hunk 5]
```diff
@@ -149,3 +149,70 @@ def test_alteration_share_classes_optional(session):
...
+    business = factory_business(identifier)
+
+    f = copy.deepcopy(ALTERATION_FILING_TEMPLATE)
+    f['filing']['header']['identifier'] = identifier
+    del f['filing']['alteration']['nameRequest']
+
+    if memorandumFileKey:
+        f['filing']['alteration']['rulesFileKey'] = memorandumFileKey
+    if memorandumFileName:
+        f['filing']['alteration']['rulesFileName'] = memorandumFileName
```
Comment (argush3) [inline legal-api/tests/unit/services/filings/validations/test_alteration.py:205]: referencing rules related keys.  please update to use memorandum keys

File: legal-api/src/legal_api/services/filings/validations/alteration.py [hunk 6]
```diff
@@ -120,3 +122,31 @@ def type_change_validation(filing):
...
+def memorandum_change_validation(filing):
+    """Validate memorandum change."""
+    msg = []
+    memorandum_file_key: Final = get_str(filing, '/filing/alteration/memorandumFileKey')
+    memorandum_file_name: Final = get_str(filing, '/filing/alteration/memorandumFileName')
+
+    if memorandum_file_key or memorandum_file_name:
+        if not (memorandum_file_key and memorandum_file_name):
+            msg.append({'error': babel('Both memorandumFileKey and memorandumFileName should be provided'),
+                        'path': '/filing/alteration/memorandumFileKey and /filing/alteration/memorandumFileName'})
```
Comment (argush3) [inline legal-api/src/legal_api/services/filings/validations/alteration.py:150]: can we provide the path for the specific memorandum property missing?  same for rules above

**Review by argush3 — APPROVED** [review]

**Review by vysakh-menon-aot — APPROVED** [review]

**Pull request merged**